Ability to alias "pluck_attributes" column names

jzaleski opened the pull request:
This new feature adds the ability to alias the keys (column names) of
the results returned from "pluck_attributes" (while still maintaining
the proper type-casts).
- Add tests for "pluck_attributes" to cover existing functionality
- Add tests for "pluck_attributes" to cover new alias[ing] functionality
- Minor clean-up/refactoring
- Update README.md

jzaleski committed:
Ability to alias "pluck_attributes" column names

This new feature adds the ability to alias the keys (column names) of
the results returned from "pluck_attributes" (while still maintaining
the proper type-casts).

* Add tests for "pluck_attributes" to cover existing functionality
* Add tests for "pluck_attributes" to cover new alias[ing] functionality
* Minor clean-up/refactoring
* Update README.md

jzaleski commented:
@dimko this is the feature I was looking to get in. I will rework https://github.com/dimko/pluckeroid/pull/3 if you end up merging this first (or vice-versa).

Hope all is well!

jzaleski commented:
@dimko I'm assuming you're probably very busy w/ work & other projects (and I totally understand!). Do you think you'll have a chance to review this PR and the other sometime soon? If not I will fork into my company repo an add the changes there. Please advise.

dmeremyanin commented on lib/pluckeroid/active_record/relation.rb:
```diff
@@ -14,11 +14,23 @@ module Relation
+            attributes.each do |key, _|
+              mapped_attributes[column_mappings[key] || key] = \
+                klass.type_cast_attribute(key, attributes)
+            end
```
Do you think we really need to use a hash with indifferent access here?

jzaleski replied:
I prefer to use symbols for keys. By using a `HashWithIndifferentAccess` it still keeps the interface consistent for people already using a previous version of this gem and gives the flexibility to use symbols to address values (as well) w/ minimal overhead.

The more I think about it, since a user of the gem can `pluck` a value or attributes by a string or a symbol, I think it is perfectly reasonable to allow for the same behavior on the result.

dmeremyanin commented on lib/pluckeroid/active_record/relation.rb:
```diff
@@ -73,6 +85,16 @@ def pluck_columns(column_names)
+      end
+
+      def extract_column_mappings(options)
+        (options[:column_mappings] || {}).with_indifferent_access
```
Here we can use `stringify_keys` instead of indifferent access

jzaleski replied:
That's fair. I made the change.

jzaleski committed:
Use `stringify_keys` instead of `with_indifferent_access`

Since the `attributes` on models are keyed by strings, it is fine to
simply stringify the keys of the `column_mappings`

dmeremyanin commented:
I'm thinking about performance:
- one more hash allocation in each iteration
- HWIA is really slower than native hash

dmeremyanin commented:
Benchmark: https://gist.github.com/dimko/08d12f9a9daad5ddac11

```
Calculating -------------------------------------
            original        17 i/100ms
             mapping        14 i/100ms
                hwia        11 i/100ms
        hwia+mapping         8 i/100ms
-------------------------------------------------
            original      200.4 (±11.0%) i/s -       986 in   5.001664s
             mapping      166.2 (±4.8%) i/s -        840 in   5.067277s
                hwia      124.8 (±7.2%) i/s -        627 in   5.050656s
        hwia+mapping      118.7 (±5.9%) i/s -        592 in   5.003750s
```

jzaleski committed:
Incorporating feedback / performance tweaks

* No longer use `HashWithIndiffentAccess` for mapped results (the
  performance hit is not worth it)
* Break out result type-casting into 2 paths, one that only handles the
  type-casting (previous behavior) and another that will apply mappings,
  if any are defined (new behavior)

jzaleski commented:
@dimko thank you for the feedback. I'm glad to see someone serious about performance. After doing some benchmarking of my own I made a few tweaks to the code for this pull-request.

With these additions the default behavior remains speedy (by skipping the lookups entirely if the `column_mappings` are falsy).

Beyond that, it is clear that `HashWithIndifferentAccess` was the wrong data-structure for the mapping (we're using a plain `Hash` now).

```
Calculating -------------------------------------
     without_mapping        18 i/100ms
        with_mapping        16 i/100ms
   with_mapping_hwia        13 i/100ms
-------------------------------------------------
     without_mapping      193.6 (±2.1%) i/s -        972 in   5.023291s
        with_mapping      166.4 (±9.0%) i/s -        832 in   5.051210s
   with_mapping_hwia      134.2 (±8.9%) i/s -        676 in   5.087792s
```

**Results based on this gist:**

https://gist.github.com/jzaleski/51ba21804386fa96257c

Looking forward to your continued feedback!

jzaleski commented:
Just circling back.. Is there anything else I can do to help get this merged at this point?
